### PR TITLE
[refactor] Let the matrix argument be compiled as a struct

### PR DESCRIPTION
**Pull request by lin-hitonami**
### Summary
🧮🚀🛠️

This pull request adds support for typed tensors as a new argument type for kernels in the C and C++ APIs of Taichi. It updates the API reference, documentation, graph representation, and test cases to use the new way of setting vector and matrix arguments as tensors with element shapes and indices. It also simplifies and enhances the Python API for matrix arguments.

> _`Tensor` is the new type of doom_
> _Binding arguments to the kernel's womb_
> _`Matrix` and `Vector` are now one_
> _Setting their shapes with `TiDataType` and `TiTensor`_

### Walkthrough
*  Add a new argument type `TI_ARGUMENT_TYPE_TENSOR` and a new union member `tensor` to support typed tensors as kernel arguments ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-4f81eb42782d0a66ae9c536581b6f052c7108a2ddf16579860f8288a5bbbd503R466-R467), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-4f81eb42782d0a66ae9c536581b6f052c7108a2ddf16579860f8288a5bbbd503R853-R854), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-693e7dacdac48ac69a749431b9e09ec9789a5aa2d4a98a4a3dad596c1fda6abbR795-R816), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-dba7c9488489c350e8d3d798f64036bc4d1e4470a75f0646e428982bcfb4cf19L159-R160), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-dba7c9488489c350e8d3d798f64036bc4d1e4470a75f0646e428982bcfb4cf19R566-R569), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-72d3acf75ae99e2cd7fcc0520273c959d0736661752c1ac1070f225e20ff213dR501), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-72d3acf75ae99e2cd7fcc0520273c959d0736661752c1ac1070f225e20ff213dR932))
*  Add new types `TiTensorValue`, `TiTensorValueWithLength`, and `TiTensor` to represent typed tensors and their values in the C API and the JSON schema ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-4f81eb42782d0a66ae9c536581b6f052c7108a2ddf16579860f8288a5bbbd503R807-R836), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-dba7c9488489c350e8d3d798f64036bc4d1e4470a75f0646e428982bcfb4cf19R489-R542))
*  Update the `TI_C_API_VERSION` macro and the API reference documentation in `taichi_core.h`, `taichi_core.h.md`, and `taichi_core.md` to reflect the new API changes ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-4f81eb42782d0a66ae9c536581b6f052c7108a2ddf16579860f8288a5bbbd503L230-R230), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-9e3473989a57d296a446c9da7fdacdc5af1f076daaf827baf7925c29d1099bb2R308), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-9e3473989a57d296a446c9da7fdacdc5af1f076daaf827baf7925c29d1099bb2R454-R470), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-9e3473989a57d296a446c9da7fdacdc5af1f076daaf827baf7925c29d1099bb2R480), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-72d3acf75ae99e2cd7fcc0520273c959d0736661752c1ac1070f225e20ff213dR513), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-72d3acf75ae99e2cd7fcc0520273c959d0736661752c1ac1070f225e20ff213dR943))
*  Add a template struct `DataTypeToEnum` and its specializations to `taichi.hpp` to provide a compile-time mapping from C++ types to `TiDataType` enumeration values ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-15e4c7e6e43328c5d3d5d3bf87f54fac89fb09ba2158bc9cd34a398f06551cf7L840-R859))
*  Modify the `push_arg` function in the `Kernel` class in `taichi.hpp` to handle vector arguments as tensors ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-15e4c7e6e43328c5d3d5d3bf87f54fac89fb09ba2158bc9cd34a398f06551cf7L887-R911))
*  Modify the `PER_BUFFER_FORMAT` macro in `export_lang.cpp` to handle matrix arguments as tensors ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-af631a0c71978fe591e17005f01f7c06bc30ae36c65df306bbb3b08ade770167L620-R669))
*  Modify the `jit_run` method of `CompiledGraph` in `graph_data.cpp` to handle matrix arguments as tensors ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-39cf507495ad085f8581e2bcd77f65188014888dde181c451ac0962166be548dL92-R127))
*  Modify the `run_kernel_test1` and `run_kernel_test2` functions in `kernel_aot_test.cpp` and `gfx_utils.cpp` to use the `set_struct_arg` method for setting vector arguments ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-39cf507495ad085f8581e2bcd77f65188014888dde181c451ac0962166be548dR47-R81), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-87c49ccfc97d474ac7265446b70631c6498ac144ca53c3258be487c9c083f7b6L175-R175), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-2fcdc6f799e8b0b12b7dd00e5de176c8ef742b5b83dfa186ca8a0ef721004137L57-R57), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-2fcdc6f799e8b0b12b7dd00e5de176c8ef742b5b83dfa186ca8a0ef721004137L103-R103))
*  Modify the `_make_arg_matrix` function in `_graph.py` to return a single `Arg` object with `ArgKind.MATRIX` and the element shape for matrix arguments ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-3b3dff33121c57d33ae7fd0a1288da627f08baed4cdb4039882cc397a2107c0cL238-R238))
*  Modify the `decl_matrix_arg` and `func__` functions in `kernel_arguments.py` and `kernel_impl.py` to use the new `set_kernel_struct_args` method of `MatrixType` for setting matrix arguments ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-575efc738df7b1202370c2531ec82232dc7f287b2bec4999af03ef40da4f5deeL85-R88), [link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-a157043b38542c8145447ff342fda65fe4d54fb777fe514daa70007e83e20dc1L772-R793))
*  Modify the `produce_injected_args` function in `utils.py` to use the `element_shape` attribute of a symbolic argument with `MatrixType` ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-6ce6635d9d3b5a4add63601f3bcbc5fb59d702b4b3b1a6685c22888257e99e3aL100-R102))
*  Remove the unused import of `VectorType` and `make_matrix` from `kernel_arguments.py` ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-575efc738df7b1202370c2531ec82232dc7f287b2bec4999af03ef40da4f5deeL10-R10))
*  Remove the outdated test case `test_exceed_max_64` from `test_argument.py` ([link](https://github.com/taichi-dev/taichi/pull/7892/files?diff=unified&w=0#diff-f2d9c4a8c504d2073549cde99b31a733a2f39bba4ce9f062cbaa09e1be17aa87L7-L32))

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #7892

### CODE REVIEW

File: python/taichi/lang/kernel_arguments.py [hunk 1]
```diff
@@ -82,14 +82,10 @@ def get_type_for_kernel_args(dtype, name):
...
-    if isinstance(matrixtype, VectorType):
-        return make_matrix([decl_scalar_arg(matrixtype.dtype, f"{name}_{i}") for i in range(matrixtype.n)])
-    return make_matrix(
-        [
-            [decl_scalar_arg(matrixtype.dtype, f"{name}_{i}_{j}") for i in range(matrixtype.m)]
-            for j in range(matrixtype.n)
-        ]
-    )
+    arg_type = get_type_for_kernel_args(matrixtype, name)
+    arg_id = impl.get_runtime().compiling_callable.insert_scalar_param(arg_type, name)
```
Comment (ailzhang) [inline python/taichi/lang/kernel_arguments.py:86]: nit: are we going to rename `insert_scalar_param` in the future PRs?

**Review by ailzhang — APPROVED** [review]
Nice!!

File: python/taichi/lang/kernel_arguments.py (hunk 1 above)
Comment (lin-hitonami) [reply python/taichi/lang/kernel_arguments.py:86]: Maybe we can merge `insert_scalar_param` and `insert_pointer_param` to a `insert_param` with an option `is_ptr`

**Pull request merged**